diff --git a/sieve/fb.cpp b/sieve/fb.cpp
index 99a66ce2b..120099f67 100644
--- a/sieve/fb.cpp
+++ b/sieve/fb.cpp
@@ -1707,7 +1707,7 @@ int fb_factorbase::read(char const * const filename)
     FILE * fbfile;
     // too small linesize led to a problem with rsa768;
     // it would probably be a good idea to get rid of fgets
-    size_t const linesize = 1000;
+    size_t const linesize = 2000;
     char line[linesize];
     unsigned long linenr = 0;
     fbprime_t maxprime = 0;
